<launch>
    <param name="/use_sim_time" value="true" />

    <!--Playing ROS bag w/ tf, must be together with others-->
    <arg name="bag_filenames"/>
    <node name="playbag" pkg="rosbag" type="play" args="--clock $(arg bag_filenames) /tf:=/tf_rm" />

    <!-- Launch Cartographer slam -->
    <arg name="lua_filename" default="mit_stata_2d.lua"/>
    <node name="cartographer_node" pkg="cartographer_ros"
        type="cartographer_node" args="
            -configuration_directory $(find cartographer_ros)/configuration_files
            -configuration_basename $(arg lua_filename)"
        output="screen">
        <remap from="scan" to="/base_scan" />
        <remap from="odom"     to="/base_odometry/odom_footprint"/>
        <remap from="imu"     to="/torso_lift_imu/data"/>
    </node>
    <!--Launch cartographer grid map transformer-->
    <node name="cartographer_occupancy_grid_node" pkg="cartographer_ros"
        type="cartographer_occupancy_grid_node" args="-resolution 0.05" output="screen">
    </node>

    <!--Corrected odometry-->
    <node name="odometry_add_frameid" pkg="cartographer_ros" type="odometry_add_frameid.py"/>

    <!--Corrected tf odom to base_footprint-->
    <node name="tf_remove_transform" pkg="cartographer_ros" type="tf_remove_transform.py"/>

    <!--point cloud from depth map and rgb image-->
    <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager"/>
    <node pkg="nodelet" type="nodelet" name="depth_image_proc" args="load depth_image_proc/point_cloud_xyzrgb standalone_nodelet" output="screen">
        <remap from="rgb/camera_info"     to="/camera/rgb/camera_info"/>
        <remap from="rgb/image_rect_color"     to="/camera/rgb/image_raw"/>
        <remap from="depth_registered/image_rect"     to="/camera/depth/image_raw"/>
        <remap from="depth_registered/points"     to="/camera/depth_registered/points"/>
    </node>

    <!--3D point registering-->
	<node pkg="octomap_server" type="octomap_server_node" name="octomap_server">
		<param name="resolution" value="0.05" />
		<!-- fixed map frame (set to 'map' if SLAM or localization running!) -->
		<param name="frame_id" type="string" value="map" />		
		<!-- maximum range to integrate (speedup!) -->
		<param name="sensor_model/max_range" value="5.0" />
		<!-- data source to integrate (PointCloud2)  move_group/filtered_cloud -->
		<remap from="cloud_in" to="camera/depth_registered/points" />
		<!--remap from="cloud_in" to="/basecam/depth_registered/points" /-->	 
		<param name="pointcloud_min_z" value="0.05" />
		<param name="pointcloud_max_z" value="2.5" />
		<param name="occupancy_min_z" value="0.05" />
		<param name="occupancy_max_z" value="2.5" />
	</node>

    <node name="rviz" pkg="rviz" type="rviz" required="true"
        args="-d $(find cartographer_ros)/configuration_files/demo_2d.rviz" />

</launch>
